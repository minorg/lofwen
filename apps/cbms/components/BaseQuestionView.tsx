import type { PropsWithChildren } from "react";
import { View } from "react-native";
import { Text } from "~/components/ui/text";
import type { Question } from "~/models/Question";

export function BaseQuestionView({
  children,
  question,
}: PropsWithChildren<{ question: Question }>) {
  return (
    <View className="flex flex-col gap-2 native:justify-center native:p-4">
      <Text className="text-primary">{question.prompt}</Text>
      {children}
    </View>
  );
}
