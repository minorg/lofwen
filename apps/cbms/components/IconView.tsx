import {
  FontAwesome6,
  MaterialCommunityIcons,
  MaterialIcons,
} from "@expo/vector-icons";
import type { OpaqueColorValue } from "react-native";
import type { Icon } from "~/models/Icon";

export function IconView({
  icon,
  ...props
}: { color?: string | OpaqueColorValue; icon: Icon; size?: number }) {
  switch (icon.family) {
    case "FontAwesome6":
      return <FontAwesome6 name={icon.name} {...props} />;
    case "MaterialCommunityIcons":
      return <MaterialCommunityIcons name={icon.name} {...props} />;
    case "MaterialIcons":
      return <MaterialIcons name={icon.name} {...props} />;
  }
}
