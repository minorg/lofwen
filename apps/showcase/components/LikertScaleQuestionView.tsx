import type { LikertScaleAnswer, LikertScaleQuestion } from "@lofwen/models";
import { Label } from "@lofwen/ui/src/components/ui/label";
import {
  RadioGroup,
  RadioGroupItem,
} from "@lofwen/ui/src/components/ui/radio-group";
import { Text } from "@lofwen/ui/src/components/ui/text";
import { useCallback } from "react";
import { View } from "react-native";

export function LikertScaleQuestionView({
  answer,
  onAnswer,
  question,
}: {
  answer: LikertScaleAnswer | null;
  onAnswer: (answer: LikertScaleAnswer) => void;
  question: LikertScaleQuestion;
}) {
  const onSelectResponseCategoryLabel = useCallback(
    (responseCategoryLabel: string) =>
      onAnswer({
        "@type": "LikertScaleAnswer",
        questionId: question["@id"],
        responseCategory: question.responseCategories.find(
          (responseCategory) =>
            responseCategory.label === responseCategoryLabel,
        )!,
      }),
    [onAnswer, question],
  );

  return (
    <View className="flex flex-col gap-2 native:justify-center native:px-4">
      <Text className="text-2xl">{question.prompt}</Text>
      <RadioGroup
        onValueChange={onSelectResponseCategoryLabel}
        value={answer?.responseCategory.label ?? ""}
      >
        {question.responseCategories.map((responseCategory) => (
          <View
            className={"flex-row items-center gap-2"}
            key={responseCategory.label}
          >
            <RadioGroupItem
              aria-labelledby={`label-for-${responseCategory.label}`}
              value={responseCategory.label}
            />
            <Label
              nativeID={`label-for-${responseCategory.label}`}
              onPress={() =>
                onSelectResponseCategoryLabel(responseCategory.label)
              }
            >
              <Text className="text-xl">{responseCategory.label}</Text>
            </Label>
          </View>
        ))}
      </RadioGroup>
    </View>
  );
}
