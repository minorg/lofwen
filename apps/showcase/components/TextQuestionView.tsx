import type { TextAnswer, TextQuestion } from "@lofwen/models";
import { Button } from "@lofwen/ui/src/components/ui/button";
import { Input } from "@lofwen/ui/src/components/ui/input";
import { Text } from "@lofwen/ui/src/components/ui/text";
import { useCallback, useEffect, useState } from "react";
import { View } from "react-native";

export function TextQuestionView({
  answer,
  onAnswer,
  question,
}: {
  answer: TextAnswer | null;
  onAnswer: (answer: TextAnswer) => void;
  question: TextQuestion;
}) {
  const [text, setText] = useState("");

  const onSubmitButtonPress = useCallback(
    () =>
      onAnswer({
        questionId: question["@id"],
        text,
        "@type": "TextAnswer",
      } satisfies TextAnswer),
    [onAnswer, question, text],
  );

  useEffect(() => {
    if (answer !== null) {
      setText(answer.text);
    }
  }, [answer]);

  return (
    <View className="flex flex-col flex-1 gap-2 native:justify-center native:px-4">
      <Text className="text-2xl">{question.prompt}</Text>
      <Input
        className="h-[8rem]"
        multiline
        onChangeText={setText}
        value={text}
      />
      <Button
        disabled={text.length === 0}
        onPress={onSubmitButtonPress}
        variant="outline"
      >
        <Text>Submit</Text>
      </Button>
    </View>
  );
}
