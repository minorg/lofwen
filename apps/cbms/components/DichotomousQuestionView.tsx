import type { DichotomousAnswer, DichotomousQuestion } from "@lofwen/models";
import { useCallback } from "react";
import { View } from "react-native";
import { BaseQuestionView } from "~/components/BaseQuestionView";
import { Label } from "~/components/ui/label";
import { RadioGroup, RadioGroupItem } from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";

export function DichotomousQuestionView({
  answer,
  onAnswer,
  question,
}: {
  answer: DichotomousAnswer | null;
  onAnswer: (answer: DichotomousAnswer) => void;
  question: DichotomousQuestion;
}) {
  const onSelectResponseCategoryLabel = useCallback(
    (responseCategoryLabel: string) =>
      onAnswer({
        "@type": "DichotomousAnswer",
        responseCategory: question.responseCategories.find(
          (responseCategory) =>
            responseCategory.label === responseCategoryLabel,
        )!,
      }),
    [onAnswer, question],
  );

  return (
    <BaseQuestionView question={question}>
      <RadioGroup
        onValueChange={onSelectResponseCategoryLabel}
        value={answer?.responseCategory.label ?? ""}
      >
        {question.responseCategories.map((responseCategory) => (
          <View
            className={"flex-row items-center gap-2"}
            key={responseCategory.label}
          >
            <RadioGroupItem
              aria-labelledby={`label-for-${responseCategory.label}`}
              value={responseCategory.label}
            />
            <Label
              nativeID={`label-for-${responseCategory.label}`}
              onPress={() =>
                onSelectResponseCategoryLabel(responseCategory.label)
              }
            >
              <Text className="text-secondary">{responseCategory.label}</Text>
            </Label>
          </View>
        ))}
      </RadioGroup>
    </BaseQuestionView>
  );
}
