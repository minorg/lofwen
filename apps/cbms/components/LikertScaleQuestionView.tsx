import type { LikertScaleAnswer, LikertScaleQuestion } from "@lofwen/models";
import { useCallback } from "react";
import { View } from "react-native";
import { BaseQuestionView } from "~/components/BaseQuestionView";
import { Label } from "~/components/ui/label";
import { RadioGroup, RadioGroupItem } from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";

export function LikertScaleQuestionView({
  answer,
  onAnswer,
  question,
}: {
  answer: LikertScaleAnswer | null;
  onAnswer: (answer: LikertScaleAnswer) => void;
  question: LikertScaleQuestion;
}) {
  const onSelectResponseCategoryLabel = useCallback(
    (responseCategoryLabel: string) =>
      onAnswer({
        "@type": "LikertScaleAnswer",
        responseCategory: question.responseCategories.find(
          (responseCategory) =>
            responseCategory.label === responseCategoryLabel,
        )!,
      }),
    [onAnswer, question],
  );

  return (
    <BaseQuestionView question={question}>
      <RadioGroup
        onValueChange={onSelectResponseCategoryLabel}
        value={answer?.responseCategory.label ?? ""}
      >
        {question.responseCategories.map((responseCategory) => (
          <View
            className={"flex-row items-center gap-2"}
            key={responseCategory.label}
          >
            <RadioGroupItem
              aria-labelledby={`label-for-${responseCategory.label}`}
              value={responseCategory.label}
            />
            <Label
              nativeID={`label-for-${responseCategory.label}`}
              onPress={() =>
                onSelectResponseCategoryLabel(responseCategory.label)
              }
            >
              <Text className="text-secondary">{responseCategory.label}</Text>
            </Label>
          </View>
        ))}
      </RadioGroup>
    </BaseQuestionView>
  );
}
