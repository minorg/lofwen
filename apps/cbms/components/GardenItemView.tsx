import { Pressable, View } from "react-native";
import { IconView } from "~/components/IconView";
import { Button } from "~/components/ui/button";
import { Text } from "~/components/ui/text";
import { useColorScheme } from "~/hooks/useColorScheme";
import type { GardenItem } from "~/models/GardenItem";

export function GardenItemView({
  gardenItem,
  onSelect,
}: { gardenItem: GardenItem; onSelect: () => void }) {
  const { isDarkColorScheme } = useColorScheme();

  return (
    <View className="border-2 border-primary flex flex-row gap-4 p-4 rounded-sm">
      <View className="flex flex-col justify-center">
        <Pressable onPress={onSelect}>
          <IconView
            color={isDarkColorScheme ? "white" : undefined}
            icon={gardenItem.icon}
            size={64}
          />
        </Pressable>
      </View>
      <View className="flex flex-col flex-1 items-end">
        <Button onPress={onSelect} variant="outline">
          <Text className="text-lg font-bold">{gardenItem.title}</Text>
        </Button>
        <Text className="text-secondary">{gardenItem.text}</Text>
      </View>
    </View>
  );
}
