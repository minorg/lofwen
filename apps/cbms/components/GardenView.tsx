import { ImageBackground, View } from "react-native";
import { GardenItemView } from "~/components/GardenItemView";
import type { Garden } from "~/models/Garden";
import type { GardenItem } from "~/models/GardenItem";

export function GardenView({
  garden,
  onSelectItem,
}: { garden: Garden; onSelectItem: (item: GardenItem) => void }) {
  return (
    <ImageBackground
      className="flex-1"
      imageClassName="opacity-10"
      source={require("../assets/images/morris-african-marigold.jpg")}
    >
      <View className="flex flex-col flex-1 gap-4 opacity-100 native:p-4">
        {garden.items.map((item) => (
          <GardenItemView
            gardenItem={item}
            key={item["@id"]}
            onSelect={() => onSelectItem(item)}
          />
        ))}
      </View>
    </ImageBackground>
  );
}
