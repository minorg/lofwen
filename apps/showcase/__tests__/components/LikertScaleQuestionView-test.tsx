import { render } from "@testing-library/react-native";
// biome-ignore lint/correctness/noUnusedImports: <explanation>
import React from "react";
import { LikertScaleQuestionView } from "~/components/LikertScaleQuestionView";
import { likertScaleQuestion } from "../data";

it("renders correctly", () => {
  const tree = render(
    <LikertScaleQuestionView
      answer={null}
      onAnswer={() => {}}
      question={likertScaleQuestion}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
