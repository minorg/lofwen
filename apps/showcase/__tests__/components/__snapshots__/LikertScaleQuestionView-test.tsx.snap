// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  className="flex flex-col gap-2 native:justify-center native:px-4"
>
  <Text
    className="text-foreground web:select-text text-2xl"
  >
    Is this the best app ever?
  </Text>
  <View
    className="web:grid gap-2"
    role="radiogroup"
  >
    <View
      className="flex-row items-center gap-2"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-labelledby="label-for-Strongly disagree"
        className="aspect-square h-4 w-4 native:h-5 native:w-5 rounded-full justify-center items-center border border-primary text-primary web:ring-offset-background web:focus:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="radio"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="web:cursor-default"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          className="text-sm text-foreground native:text-base font-medium leading-none web:peer-disabled:cursor-not-allowed web:peer-disabled:opacity-70"
          nativeID="label-for-Strongly disagree"
        >
          <Text
            className="text-foreground web:select-text text-xl"
          >
            Strongly disagree
          </Text>
        </Text>
      </View>
    </View>
    <View
      className="flex-row items-center gap-2"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-labelledby="label-for-Disagree"
        className="aspect-square h-4 w-4 native:h-5 native:w-5 rounded-full justify-center items-center border border-primary text-primary web:ring-offset-background web:focus:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="radio"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="web:cursor-default"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          className="text-sm text-foreground native:text-base font-medium leading-none web:peer-disabled:cursor-not-allowed web:peer-disabled:opacity-70"
          nativeID="label-for-Disagree"
        >
          <Text
            className="text-foreground web:select-text text-xl"
          >
            Disagree
          </Text>
        </Text>
      </View>
    </View>
    <View
      className="flex-row items-center gap-2"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-labelledby="label-for-Neutral"
        className="aspect-square h-4 w-4 native:h-5 native:w-5 rounded-full justify-center items-center border border-primary text-primary web:ring-offset-background web:focus:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="radio"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="web:cursor-default"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          className="text-sm text-foreground native:text-base font-medium leading-none web:peer-disabled:cursor-not-allowed web:peer-disabled:opacity-70"
          nativeID="label-for-Neutral"
        >
          <Text
            className="text-foreground web:select-text text-xl"
          >
            Neutral
          </Text>
        </Text>
      </View>
    </View>
    <View
      className="flex-row items-center gap-2"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-labelledby="label-for-Agree"
        className="aspect-square h-4 w-4 native:h-5 native:w-5 rounded-full justify-center items-center border border-primary text-primary web:ring-offset-background web:focus:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="radio"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="web:cursor-default"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          className="text-sm text-foreground native:text-base font-medium leading-none web:peer-disabled:cursor-not-allowed web:peer-disabled:opacity-70"
          nativeID="label-for-Agree"
        >
          <Text
            className="text-foreground web:select-text text-xl"
          >
            Agree
          </Text>
        </Text>
      </View>
    </View>
    <View
      className="flex-row items-center gap-2"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        aria-labelledby="label-for-Strongly agree"
        className="aspect-square h-4 w-4 native:h-5 native:w-5 rounded-full justify-center items-center border border-primary text-primary web:ring-offset-background web:focus:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="radio"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="web:cursor-default"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          className="text-sm text-foreground native:text-base font-medium leading-none web:peer-disabled:cursor-not-allowed web:peer-disabled:opacity-70"
          nativeID="label-for-Strongly agree"
        >
          <Text
            className="text-foreground web:select-text text-xl"
          >
            Strongly agree
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
